### ISSUE TITLE
Deploy SpeciesNet

### ISSUE THREAD

**Issue by nathanielrindlaub** [tnc-ca-geo/animl-ml#126]
Wildlife Insights and Google have open-sourced SpeciesNet (!), the ~2000 class classifier that WI uses in its app. The [documentation](https://github.com/google/cameratrapai/) looks superb, and they even have some [server](https://github.com/google/cameratrapai/blob/main/speciesnet/scripts/run_server.py) code for standing up a REST API wrapper around it. We should deploy this to Animl ASAP as it would provide a ton of value to users who don't have a more localized classifier for their area of study.

A couple of implementation notes:
- The repo includes an [ensemble pipeline](https://github.com/google/cameratrapai?tab=readme-ov-file#ensemble-decision-making) that requests detections from MegaDetector before passing the resulting crops on to the SpeciesNet classifier for species predictions. We do not need that. In Animl, users configure and define the sequencing of their ML models using [Automation Rules](https://docs.animl.camera/fundamentals/automation-rules), so we can assume that like other classifiers, users will have first configured an Automation Rule to request detections and bounding boxes from MegaDetector, and they will have configured a second Automation Rule to request predictions from SpeciesNet when an "animal" is detected, using the bounding boxes provided by MegaDetector. So we only really need the `always_crop` [model](https://github.com/google/cameratrapai?tab=readme-ov-file#supported-models).
- that said, there are some post processing steps that are part of the ensemble pipeline that would be cool to replicate in Animl. Specifically, the geofencing and label rollup. To implement geofencing we'd have to provide the user with some additional forms fields when they're configuring their SpeciesNet AutomationRules so they can enter their [ISO 3166-1 alpha-3 three-letter country code](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-3) and if they are in the USA, a two-letter state abbreviation.
- The label rollup process I don't fully understand yet and I think is less of a priority. Once we have deployed the model and experimented with it and gotten a better feel for its output, I think we may want to consider implementing the rollup logic, but for now let's put a pin in that.
- I have seen an optional parameter for `target_species` in some of the scripts (e.g. [here](https://github.com/google/cameratrapai/blob/a47bfac3407c58c724e41eb90a16241fbe3dcbf8/speciesnet/classifier.py#L50) in `classifier.py`). I don't know exactly what this does yet, but if it provides an additional level of targeting/filtering, perhaps that's an option we should expose to user too.

I propose we deploy the model in the following phases:

### Phase 1:  Containerize and deploy the model to Sagemaker Serverless Endpoint
1. **Containerize the SpeciesNet classifier.** The approach I'd try first is to containerize the LitServe-based [server]([server](https://github.com/google/cameratrapai/blob/main/speciesnet/scripts/run_server.py) ) code. LitServe has [tooling](https://lightning.ai/docs/litserve/features/dockerization-deployment) to dockerize LitServe apps so that might be all we need. Note: I assume that the `run_server.py` code also includes the ability to run the full ensemble pipeline, including MegaDetector, which again we don't really need, so there may be opportunities to reduce bloat by stripping it down to the classifier only. If we strike out containerizing the model via LitServe, we may need to use TensorFlow Serving instead (SpeciesNet is a TF/Keras model).
2. **Deploy as Sagemaker Serverless Endpoint.** The [`mdv5_deploy.ipynb`](https://github.com/tnc-ca-geo/animl-ml/blob/master/api/megadetectorv5/mdv5_deploy.ipynb) in this repo outlines the core steps required.

### Phase 2: Integrate endpoint with Animl
1. **Script the creation of JSON [`MLModel.categories`](https://github.com/tnc-ca-geo/animl-api/blob/4f673ca1c9544ec4e4166f13a899ee652f276422/src/api/db/schemas/MLModel.ts#L32-L36) from SpeciesNet's labels** - once you download the model files the labels can be found in `always_crop_99710272_22x8_v12_epoch_00148.labels.txt`. We will need to create a script that iterates over each of the labels in the `.txt` file (which look like: `8a57d557-41a8-4282-9e37-59f1b9e806ea;mammalia;primates;callitrichidae;saguinus;melanoleucus;white/crandall's saddleback tamarin`) and creates an array of JSON objects with the shape `{ _id: "8a57d557-41a8-4282-9e37-59f1b9e806ea", name: "white/crandall's saddleback tamarin", color: "#E54D2E"}`. A few notes about the script:
    - in addition to the ID and common name, SpeciesNet species labels include the full taxonomic classification: i.e. the `...mammalia;primates;callitrichidae;saguinus;melanoleucus;...` part. We will have to decide what to do all of that information. From the UI perspective, we really only need the common names (and IDs), but if we implement the label rollup post processing step described above this is likely necessary to retain somehow. For an MVP implementation, we can just move forward with the current `MLModel.category` schema and ignore the full taxonomic classification, but I think we should consider adding an optional `taxonomy` field to at least retain that information.
    - the colors can be assigned by looping over the default Label colors (see Nick's [Gist](https://gist.github.com/ingalls/7b7325688a6fd24e9c49cc0992530043) from a prior DB migration)
    - I recently discovered the `MLModel.categories` `name` fields cannot contain periods and dollar signs (more info on that [here](https://github.com/tnc-ca-geo/animl-api/issues/314)), so we will need to scrub both from the SpeciesNet common names when generating the JSON objects.
4. **Create the rest of the [`MLModel` config record](https://github.com/tnc-ca-geo/animl-api/blob/4f673ca1c9544ec4e4166f13a899ee652f276422/src/api/db/schemas/MLModel.ts#L25-L41) in MongoDB** (in the dev DB for starters). Once I have the JSON records drafted I will typically add them to the DB manually via MongoDB Compass GUI, but you could use the CLI as well. We will also need to add necessary SSM params and add a reference to the new model to `Project.availabileMLModels` for any Projects we want to test SpeciesNet on.
5. **Create the appropriate [ModelInterface](https://github.com/tnc-ca-geo/animl-api/blob/main/src/ml/modelInterfaces.ts)** in the animl-api
### Phase 3: Implement geofencing
1. **Add ability for users to include country and state codes on their SpeciesNet inference Automation Rules**
2. **Filter SpeciesNet predictions based on geofencing rules**. If we're successful in containerizing the LitServe-based `run_server.py`, this part may be done for us and would just involve passing the country/state code into the inference request.

**Comment by geohacker** [comment]
This is now complete, with the exception of SpeciesNet ensemble support #130.

* animl-ml PR https://github.com/tnc-ca-geo/animl-ml/pull/127
* animl-api https://github.com/tnc-ca-geo/animl-api/pull/315

It's deployed to the dev stack.
